# Introduction
[Reference](https://www.hackerrank.com/challenges/sherlock-and-permutations/problem?isFullScreen=true)

- Watson asks Sherlock: `Given a string` $S$ of $N$ `0's` and $M$ `1's`, `how many unique permutations of this string start with` $1$ `?`
- Print out the answer modulo (109+7).

##### Input Format
- First line contains $T$, the number of test cases.
- Each test case consists of $N$ and $M$ separated by a space.

##### Output Format
- $2 \leq T \leq 200$
- $1 \leq N, M \leq 1000$

In [1]:
import math
from itertools import permutations 

def factorial(n):
    if n == 0:
        return 1
    else:
        return (n*factorial(n-1))
def all_perm(n, m):
    return factorial(n+m) / (factorial(m) * factorial(n))

def generate_series(n, m):
    ls = n*[0] + m*[1]
    my_perm = list(permutations(ls))
    my_perm = [''.join([str(s) for s in elm]) for elm in my_perm]
    return set(my_perm)

for n, m in [(1, 1), (2, 1), (1, 2), (3, 2), (2, 3), (3, 3)]:
    print(generate_series(n, m))

{'10', '01'}
{'010', '001', '100'}
{'011', '110', '101'}
{'00011', '00101', '01010', '00110', '01100', '10010', '11000', '01001', '10100', '10001'}
{'11010', '01011', '11001', '11100', '00111', '01101', '01110', '10011', '10101', '10110'}
{'001101', '100110', '001011', '011010', '010101', '110100', '111000', '101001', '110010', '011001', '100011', '010011', '001110', '110001', '000111', '101100', '011100', '010110', '101010', '100101'}


This is easy to check that the total permutations of $(n ,m)$ position for $(0, 1)$ in a series is $$ f(n,m) := \frac{(n+m)!}{n! \times m!} $$
For instance,
- $m=n=1$ then we have `10` and `01`
- $m=2, n = 1$ then we have `110, 101` and `011`
- $m=1, n = 2$ then we have `100` and `010, 001`
- $m=2, n = 2$ then we have `1100, 1010, 1001` and `0110, 0101, 0011`
- $m=3, n = 2$ then we have `11100, 11010, 11001, 10110, 10101, 10011` and `01110, 01101, 01011, 00111`
- $m=2, n = 3$ then we have `11000, 10100, 10010, 10001` and `01100, 01010, 01001, 00110, 00101, 00011`

Now, we define 2 functions to measure the number of series which started by $1$ among ($n$ digits of $0$ and $m$ digits of $1$)

In [5]:
def solve(m, n):
    return int(factorial(n+m-1) / (factorial(n-1)*factorial(m) )) % (10**9 + 7)

def series_started_by1(m, n):
	n -= 1
	res = 1
	for i in range(n + 1, n + m + 1):
		res = res * i // (i - n)
	return res % (10**9 + 7)

for n, m in [(1, 1), (2, 1), (1, 2), (3, 2), (2, 3), (3, 3), (2, 9), (9, 2)]:
    print(f"For n={n}, m={m} we have {series_started_by1(n, m)} elements which started by 1")

For n=1, m=1 we have 1 elements which started by 1
For n=2, m=1 we have 1 elements which started by 1
For n=1, m=2 we have 2 elements which started by 1
For n=3, m=2 we have 4 elements which started by 1
For n=2, m=3 we have 6 elements which started by 1
For n=3, m=3 we have 10 elements which started by 1
For n=2, m=9 we have 45 elements which started by 1
For n=9, m=2 we have 10 elements which started by 1


The true formulation must be

$$ F(n, m) = \dfrac{(n+m-1)!}{(n-1)! \times m!} $$

The limitation of `factorial` $n!$ won't be possible to store these many digits even if we use long int, so we can re-write the equation above like this

$$ \displaystyle F(n, m) = \prod_{k=n+1}^{n+m+1} \left \lfloor \frac{k}{k - n} \right \rfloor  $$

Comparison
- Method 1 (`solve`) needs $(n+m-1) + (n-1) + m$ multiplications for 3 `factorials`
- Method 2 (`series_started_by1`) limit on range from $n+1$ to $n+m-1$

For $n, m$ large enough then method 1 will not be a good choice to solve

In [6]:
check_list = [(1, 1), (2, 1), (1, 2), (3, 3), (4, 6), (4,4), (2,4), (4,2), (3, 2), (2,3), (7, 8), (8,7), (7, 5), 
              (5, 7), (3, 3), (2, 9), (8, 3), (3,6), (6,3),
              (6, 9), (9, 2), (7, 7), (9, 9), (12, 11), (23, 20), (31, 29), (19, 81), (100, 23), (100, 90)]
for n, m in check_list:
    print(f"For n:{n} \t m:{m} \t..\t method1: {solve(n, m)} \t , method2: {series_started_by1(n, m)}")

For n:1 	 m:1 	..	 method1: 1 	 , method2: 1
For n:2 	 m:1 	..	 method1: 1 	 , method2: 1
For n:1 	 m:2 	..	 method1: 2 	 , method2: 2
For n:3 	 m:3 	..	 method1: 10 	 , method2: 10
For n:4 	 m:6 	..	 method1: 126 	 , method2: 126
For n:4 	 m:4 	..	 method1: 35 	 , method2: 35
For n:2 	 m:4 	..	 method1: 10 	 , method2: 10
For n:4 	 m:2 	..	 method1: 5 	 , method2: 5
For n:3 	 m:2 	..	 method1: 4 	 , method2: 4
For n:2 	 m:3 	..	 method1: 6 	 , method2: 6
For n:7 	 m:8 	..	 method1: 3432 	 , method2: 3432
For n:8 	 m:7 	..	 method1: 3003 	 , method2: 3003
For n:7 	 m:5 	..	 method1: 330 	 , method2: 330
For n:5 	 m:7 	..	 method1: 462 	 , method2: 462
For n:3 	 m:3 	..	 method1: 10 	 , method2: 10
For n:2 	 m:9 	..	 method1: 45 	 , method2: 45
For n:8 	 m:3 	..	 method1: 45 	 , method2: 45
For n:3 	 m:6 	..	 method1: 56 	 , method2: 56
For n:6 	 m:3 	..	 method1: 28 	 , method2: 28
For n:6 	 m:9 	..	 method1: 3003 	 , method2: 3003
For n:9 	 m:2 	..	 method1: 10 	 , method2: 10
For n:7

### Test for large number

In [4]:
check_list_large_num = [(522, 575), (772, 81), (121, 666), (982, 784), (991, 310), (999, 999)]
for n, m in check_list_large_num:
    print(series_started_by1(n, m))

769859017
586148781
360229250
597385102
154525273
482800871
